Philips.CodeAnalysis.MaintainabilityAnalyzers/Readability/AvoidMultipleLambdasOnSingleLineAnalyzer.cs

```diff
@@ -0,0 +1,87 @@
+﻿// © 2023 Koninklijke Philips N.V. See License.md in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Philips.CodeAnalysis.Common;
+
+namespace Philips.CodeAnalysis.MaintainabilityAnalyzers.Readability
+{
+	[DiagnosticAnalyzer(LanguageNames.CSharp)]
+	public class AvoidMultipleLambdasOnSingleLineAnalyzer : DiagnosticAnalyzer
+	{
+
+		private const string Title = @"Avoid putting multiple Lambda expressions on the same line";
+		public const string MessageFormat = @"Avoid putting multiple Lambda expressions on the same line";
+		private const string Description = @"Avoid putting multiple Lambda expressions on the same line. To improve readability, split them into separate lines.";
+		private const string Category = Categories.Readability;
+
+		private static readonly DiagnosticDescriptor Rule = new(Helper.ToDiagnosticId(DiagnosticIds.AvoidMultipleLambdasOnSingleLine), Title,
+			MessageFormat, Category, DiagnosticSeverity.Error, isEnabledByDefault: true, description: Description);
+
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+		private readonly HashSet<string> visitedTypes = new();
```

Comment:
HashSet isn't threadsafe. But to reduce complexity, would it make sense to only analyze if we're the first or last lambda in the statement? E.g., if ChildNodes.OfType<Lambda>.Any then skip Analyze? Or some other way to avoid needing to analyze state, similar to what we did in a different recent Analyzer?

Reply:
The relationship can be quite complex, see the the effort that the CodeFixProivider needs to do to achieve this.
I used ImmutableHashSet now and appended the path of the source code to make it (almost) unique.